### PR TITLE
Setup good-fences in tree

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Uses https://www.npmjs.com/package/good-fences to specify and enforce which dependencies are allowed within tree.

## PR Checklist

> Use the check-list below to ensure your branch is ready for PR. If the item is not applicable, leave it blank.

-   [ ] I have updated the documentation accordingly.
-   [ ] I have added tests to cover my changes.
-   [x] All new and existing tests passed.
-   [ ] My code follows the code style of this project.
-   [ ] I ran the lint checks which produced no new errors nor warnings for my changes.
-   [ ] I have checked to ensure there aren't other open Pull Requests for the same update/change.

## Does this introduce a breaking change?

-   [ ] Yes
-   [x] No

### CODE REVIEW

**Comment by CraigMacomber** [comment]
I don't know how to properly integrate this with our incremental build. As was, it warned with:

> @fluid-internal/tree: warning: non-conformant script "build"
> @fluid-internal/tree: warning:   expect: concurrently npm:build:compile npm:lint && npm run build:docs
> @fluid-internal/tree: warning:   actual: concurrently npm:build:compile npm:lint npm:fence && npm run build:docs

If I move the check into lint, it warns with `non-conformant script "lint"` instead. Similar issue if I make it part of `test`.

I suspect explicit support would have to be added to fluid-build if we want to support good-fences being run as part of the build.

I however can pragmatically invoke good-fences as a mocha test. I have made this change.

File: packages/dds/tree/package.json [hunk 1]
```diff
@@ -28,6 +28,7 @@
     "clean": "rimraf dist lib *.tsbuildinfo *.build.log",
     "eslint": "eslint --format stylish src",
     "eslint:fix": "eslint --format stylish src --fix --fix-type problem,suggestion,layout",
+    "fence": "gf",
```
Comment (noencke) [inline packages/dds/tree/package.json:31]: Do we want to `&&` this into our build script above? Otherwise I'm afraid it will not be enforced.

Comment (noencke) [reply packages/dds/tree/package.json:31]: Interesting. If this turns out to be a good experiment then maybe we can figure out how to officially put it in the build

**Review by noencke — APPROVED** [review]

**Comment by Josmithr** [comment]
It might be worth adding some simple "_comment" (or something) blocks to the fence files to explain why particular dependency edges are being enforced. Will make it easier for devs to reason about whether or not they can make changes in the future.

**Comment by CraigMacomber** [comment]
> It might be worth adding some simple "_comment" (or something) blocks to the fence files to explain why particular dependency edges are being enforced. Will make it easier for devs to reason about whether or not they can make changes in the future.
> 
> Excited to see this added to the repo though! I like this tool :)

Right now the actual fences are very much in flux: I'll be updating / clarifying the actual fences in later changes.

For documenting the fence files, it looks like it would be a trivial change to good-fences to make it support json with comments (like tsc does for its config files). I'll look into making a PR upstream for that since I really dislike adding comment fields to json.

**Pull request merged**